# Example of the Bert4Rec training and inference stages
Note that all the given examples can be run without using PySpark, using only Pandas

In [1]:
import lightning as L
from lightning.pytorch.loggers import CSVLogger
from lightning.pytorch.callbacks import ModelCheckpoint
from torch.utils.data import DataLoader
import torch

from replay.metrics import OfflineMetrics, Recall, Precision, MAP, NDCG, HitRate, MRR
from replay.metrics.torch_metrics_builder import metrics_to_df
from replay.splitters import LastNSplitter
from replay.utils import get_spark_session
from replay.data import (
    FeatureHint,
    FeatureInfo,
    FeatureSchema,
    FeatureSource,
    FeatureType,
    Dataset,
)
from replay.models.nn.optimizer_utils import FatOptimizerFactory
from replay.models.nn.sequential.callbacks import (
    ValidationMetricsCallback,
    SparkPredictionCallback,
    PandasPredictionCallback, 
    TorchPredictionCallback,
    QueryEmbeddingsPredictionCallback,
)
from replay.models.nn.sequential.postprocessors import RemoveSeenItems
from replay.data.nn import (
    SequenceTokenizer,
    SequentialDataset,
    TensorFeatureSource,
    TensorSchema,
    TensorFeatureInfo
)
from replay.models.nn.sequential import Bert4Rec
from replay.models.nn.sequential.bert4rec import (
    Bert4RecPredictionDataset,
    Bert4RecTrainingDataset,
    Bert4RecValidationDataset,
    Bert4RecPredictionBatch,
    Bert4RecModel
)

import pandas as pd

## Getting a spark session

In [ ]:
spark_session = get_spark_session()

## Prepare data
### Load raw movielens-1M interactions, item features and user features.
In the current implementation, the SASRec does not take into account the features of items or users. They are only used to get a complete list of users and items.

In [ ]:
!pip install rs-datasets

In [4]:
from rs_datasets import MovieLens

In [5]:
movielens = MovieLens("1m")
interactions = movielens.ratings
user_features = movielens.users
item_features = movielens.items

In [6]:
interactions.head()

user_id  item_id  rating  timestamp
0        1     1193       5  978300760
1        1      661       3  978302109
2        1      914       3  978301968
3        1     3408       4  978300275
4        1     2355       5  978824291

In [7]:
user_features.head()

user_id gender  age  occupation zip_code
0        1      F    1          10    48067
1        2      M   56          16    70072
2        3      M   25          15    55117
3        4      M   45           7    02460
4        5      M   25          20    55455

In [8]:
item_features.head()

item_id                               title                        genres
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy

Removing duplicates in the timestamp column without changing the original items order where timestamp is the same

In [9]:
interactions["timestamp"] = interactions["timestamp"].astype("int64")
interactions = interactions.sort_values(by="timestamp")
interactions["timestamp"] = interactions.groupby("user_id").cumcount()
interactions

user_id  item_id  rating  timestamp
1000138     6040      858       4          0
1000153     6040     2384       4          1
999873      6040      593       5          2
1000007     6040     1961       4          3
1000192     6040     2019       5          4
...          ...      ...     ...        ...
825793      4958     2399       1        446
825438      4958     1407       5        447
825724      4958     3264       4        448
825731      4958     2634       3        449
825603      4958     1924       4        450

[1000209 rows x 4 columns]

### Split interactions into the train, validation and test datasets using LastNSplitter

In [10]:
splitter = LastNSplitter(
    N=1,
    divide_column="user_id",
    query_column="user_id",
    strategy="interactions",
)

raw_test_events, raw_test_gt = splitter.split(interactions)
raw_validation_events, raw_validation_gt = splitter.split(raw_test_events)
raw_train_events = raw_validation_events

### Prepare FeatureSchema required to create Dataset

In [11]:
def prepare_feature_schema(is_ground_truth: bool) -> FeatureSchema:
    base_features = FeatureSchema(
        [
            FeatureInfo(
                column="user_id",
                feature_hint=FeatureHint.QUERY_ID,
                feature_type=FeatureType.CATEGORICAL,
            ),
            FeatureInfo(
                column="item_id",
                feature_hint=FeatureHint.ITEM_ID,
                feature_type=FeatureType.CATEGORICAL,
            ),
        ]
    )
    if is_ground_truth:
        return base_features

    all_features = base_features + FeatureSchema(
        [
            FeatureInfo(
                column="timestamp",
                feature_type=FeatureType.NUMERICAL,
                feature_hint=FeatureHint.TIMESTAMP,
            ),
        ]
    )
    return all_features

### Create Dataset for the training stage

In [12]:
train_dataset = Dataset(
    feature_schema=prepare_feature_schema(is_ground_truth=False),
    interactions=raw_train_events,
    query_features=user_features,
    item_features=item_features,
    check_consistency=True,
    categorical_encoded=False,
)

### Create Datasets (events and ground_truth) for the validation stage

In [13]:
validation_dataset = Dataset(
    feature_schema=prepare_feature_schema(is_ground_truth=False),
    interactions=raw_validation_events,
    query_features=user_features,
    item_features=item_features,
    check_consistency=True,
    categorical_encoded=False,
)
validation_gt = Dataset(
    feature_schema=prepare_feature_schema(is_ground_truth=True),
    interactions=raw_validation_gt,
    check_consistency=True,
    categorical_encoded=False,
)

### Create Datasets (events and ground_truth) for the testing stage

In [14]:
test_dataset = Dataset(
    feature_schema=prepare_feature_schema(is_ground_truth=False),
    interactions=raw_test_events,
    query_features=user_features,
    item_features=item_features,
    check_consistency=True,
    categorical_encoded=False,
)
test_gt = Dataset(
    feature_schema=prepare_feature_schema(is_ground_truth=True),
    interactions=raw_test_gt,
    check_consistency=True,
    categorical_encoded=False,
)

### Create the tensor schema
A schema shows the correspondence of columns from the source dataset with the internal representation of tensors inside the model

In [15]:
ITEM_FEATURE_NAME = "item_id_seq"

tensor_schema = TensorSchema(
    TensorFeatureInfo(
        name=ITEM_FEATURE_NAME,
        is_seq=True,
        feature_type=FeatureType.CATEGORICAL,
        feature_sources=[TensorFeatureSource(FeatureSource.INTERACTIONS, train_dataset.feature_schema.item_id_column)],
        feature_hint=FeatureHint.ITEM_ID,
        embedding_dim=300,
    )
)

### Create sequential datasets using SequenceTokenizer
The SequentialDataset internally store data in the form of sequences of items sorted by increasing interaction time (timestamp). A SequenceTokenizer is used to convert to this format. In addition, the SequenceTokenizer encodes all categorical columns from the source dataset and stores mapping inside itself.
SequentialDataset.keep_common_query_ids is used to leave only sequences from the same users

In [16]:
tokenizer = SequenceTokenizer(tensor_schema, allow_collect_to_master=True)
tokenizer.fit(train_dataset)

sequential_train_dataset = tokenizer.transform(train_dataset)

sequential_validation_dataset = tokenizer.transform(validation_dataset)
sequential_validation_gt = tokenizer.transform(validation_gt, [tensor_schema.item_id_feature_name])

sequential_validation_dataset, sequential_validation_gt = SequentialDataset.keep_common_query_ids(
    sequential_validation_dataset, sequential_validation_gt
)

In [17]:
test_query_ids = test_gt.query_ids
test_query_ids_np = tokenizer.query_id_encoder.transform(test_query_ids)["user_id"].values
sequential_test_dataset = tokenizer.transform(test_dataset).filter_by_query_id(test_query_ids_np)

You can get the user and item mapping and inverse mapping as follows

In [ ]:
print(tokenizer.query_id_encoder.mapping, tokenizer.query_id_encoder.inverse_mapping)
print(tokenizer.item_id_encoder.mapping, tokenizer.item_id_encoder.inverse_mapping)

## Train model
### Create Bert4Rec model instance and run the training stage using lightning
After each epoch validation metrics are shown. You can change the list of validation metrics in ValidationMetricsCallback
The model is determined to be the best and is saved if the metric updates its maximum during validation (see the ModelCheckpoint)

In [ ]:
MAX_SEQ_LEN = 100
BATCH_SIZE = 512
NUM_WORKERS = 4

model = Bert4Rec(
    tensor_schema,
    block_count=2,
    head_count=4,
    max_seq_len=MAX_SEQ_LEN,
    hidden_size=300,
    dropout_rate=0.5,
    optimizer_factory=FatOptimizerFactory(learning_rate=0.001),
)
checkpoint_callback = ModelCheckpoint(
    dirpath=".checkpoints",
    save_top_k=1,
    verbose=True,
    # if you use multiple dataloaders, then add the serial number of the dataloader to the suffix of the metric name.
    # For example,"recall@10/dataloader_idx_0"
    monitor="recall@10",
    mode="max",
)

validation_metrics_callback = ValidationMetricsCallback(
    metrics=["map", "ndcg", "recall"],
    ks=[1, 5, 10, 20],
    item_count=train_dataset.item_count,
    postprocessors=[RemoveSeenItems(sequential_validation_dataset)]
)

csv_logger = CSVLogger(save_dir=".logs/train", name="Bert4Rec_example")

trainer = L.Trainer(
    max_epochs=100,
    callbacks=[checkpoint_callback, validation_metrics_callback],
    logger=csv_logger,
)

train_dataloader = DataLoader(
    dataset=Bert4RecTrainingDataset(
        sequential_train_dataset,
        max_sequence_length=MAX_SEQ_LEN,
    ),
    batch_size=BATCH_SIZE,
    shuffle=True,
    num_workers=4,
    pin_memory=True,
)

validation_dataloader = DataLoader(
    dataset=Bert4RecValidationDataset(
        sequential_validation_dataset,
        sequential_validation_gt,
        sequential_train_dataset,
        max_sequence_length=MAX_SEQ_LEN,
    ),
    batch_size=BATCH_SIZE,
    num_workers=NUM_WORKERS,
    pin_memory=True,
)

trainer.fit(
    model,
    train_dataloaders=train_dataloader,
    val_dataloaders=validation_dataloader,
)

The path to the best model is saved inside checkpoint_callback

In [20]:
best_model = Bert4Rec.load_from_checkpoint(checkpoint_callback.best_model_path)

## Inference stage
### Prepare Dataloader and logger

In [21]:
prediction_dataloader = DataLoader(
    dataset=Bert4RecPredictionDataset(
        sequential_test_dataset,
        max_sequence_length=MAX_SEQ_LEN,
    ),
    batch_size=BATCH_SIZE,
    num_workers=NUM_WORKERS,
    pin_memory=True,
)

csv_logger = CSVLogger(save_dir=".logs/test", name="Bert4Rec_example")

### Run inference
You can get the recommendations in three formats: PySpark DataFrame, Pandas DataFrame, PyTorch tensors. Each of the types corresponds a callback
You can filter the results using postprocessors strategy. For example the RemoveSeenItems postprocessor is filtering out the items that already have been seen in test dataset
You don't need to use all three callbacks. This is shown only for example

Also, you can get user embeddings, that were used to perform predictions, using `get_query_embedding` method inside Bert4RecModel or `QueryEmbeddingsPredictionCallback` for lightning module.

In [ ]:
TOPK = [1, 10, 20, 100]

postprocessors = [RemoveSeenItems(sequential_test_dataset)]

spark_prediction_callback = SparkPredictionCallback(
    spark_session=spark_session,
    top_k=max(TOPK),
    query_column="user_id",
    item_column="item_id",
    rating_column="score",
    postprocessors=postprocessors,
)

pandas_prediction_callback = PandasPredictionCallback(
    top_k=max(TOPK),
    query_column="user_id",
    item_column="item_id",
    rating_column="score",
    postprocessors=postprocessors,
)

torch_prediction_callback = TorchPredictionCallback(
    top_k=max(TOPK),
    postprocessors=postprocessors,
)

query_embeddings_callback = QueryEmbeddingsPredictionCallback()

trainer = L.Trainer(
    callbacks=[
        spark_prediction_callback,
        pandas_prediction_callback,
        torch_prediction_callback,
        query_embeddings_callback,
    ], 
    logger=csv_logger, 
    inference_mode=True
)
trainer.predict(best_model, dataloaders=prediction_dataloader, return_predictions=False)

spark_res = spark_prediction_callback.get_result()
pandas_res = pandas_prediction_callback.get_result()
torch_user_ids, torch_item_ids, torch_scores = torch_prediction_callback.get_result()
user_embeddings = query_embeddings_callback.get_result()

In [23]:
spark_res.show()

+-------+-------+------------------+
|user_id|item_id|             score|
+-------+-------+------------------+
|      0|    360|12.995258331298828|
|      0|   2020|12.323760986328125|
|      0|   1642|11.679502487182617|
|      0|   2027|11.608592987060547|
|      0|     47|11.454986572265625|
|      0|   2325| 11.02070426940918|
|      0|   2014| 10.82581615447998|
|      0|     12|10.764541625976562|
|      0|   1382|10.612886428833008|
|      0|   3090|10.514969825744629|
|      0|   3144|10.471624374389648|
|      0|    547| 10.39176082611084|
|      0|   2965|10.133315086364746|
|      0|   2018|10.030275344848633|
|      0|   2021| 9.819190979003906|
|      0|   2315| 9.815552711486816|
|      0|   1019| 9.720121383666992|
|      0|   2012|    9.557373046875|
|      0|   1012|  9.46037769317627|
|      0|   2009| 9.433520317077637|
+-------+-------+------------------+
only showing top 20 rows



In [24]:
pandas_res

user_id item_id      score
0           0     360  12.995258
0           0    2020  12.323761
0           0    1642  11.679502
0           0    2027  11.608593
0           0      47  11.454987
...       ...     ...        ...
6039     6039     547   4.090594
6039     6039    1385   4.087193
6039     6039    1576   4.066721
6039     6039    1161   4.063687
6039     6039     384   4.058073

[604000 rows x 3 columns]

In [25]:
print(torch_user_ids[0], torch_item_ids[0], torch_scores[0])

tensor(0) tensor([ 360, 2020, 1642, 2027,   47, 2325, 2014,   12, 1382, 3090, 3144,  547,
        2965, 2018, 2021, 2315, 1019, 2012, 1012, 2009,  592, 2023,  974, 2016,
        2807, 2692, 2068,  700, 1002, 1020,  826, 1547,  105,  310, 3088,   59,
        2285, 3414, 1050,  582, 2070, 1010,  612, 2631, 2011, 1262,  667,  314,
        2073,  156,  259, 1850, 1664, 1205, 1990,  731,    1, 2640, 2030,  791,
        2013, 3360, 1457, 3331,   33, 3546, 2072,  626, 2830, 1346,  554, 1459,
        1979, 2033,  571,    7, 1964, 2985, 1551, 1132,  451, 1008, 2918, 2931,
        3676,  999, 1433,  236, 1017, 2047,  358, 1988, 2788, 1393, 1743, 2054,
        1011, 3685, 3329, 1983]) tensor([12.9953, 12.3238, 11.6795, 11.6086, 11.4550, 11.0207, 10.8258, 10.7645,
        10.6129, 10.5150, 10.4716, 10.3918, 10.1333, 10.0303,  9.8192,  9.8156,
         9.7201,  9.5574,  9.4604,  9.4335,  9.4125,  9.3910,  9.3880,  9.3848,
         9.2478,  9.2305,  9.1872,  9.1715,  9.1321,  9.1000,  9.0833,  9.061

Suppose we want to get the recomendations in PySpark format. 
Let's get the inverse representation of labels using inverse_transform method.

Note that the reverse representation can only be obtained for PySpark and Pandas formats. When working with PyTorch tensors, the reverse representation must be done manually

In [26]:
recommendations = tokenizer.query_and_item_id_encoder.inverse_transform(spark_res)

In [27]:
recommendations.show()

+------------------+-------+-------+
|             score|user_id|item_id|
+------------------+-------+-------+
|10.133315086364746|      1|   3034|
|10.612886428833008|      1|   1405|
| 9.433520317077637|      1|   2078|
|11.454986572265625|      1|     48|
|  9.41248607635498|      1|    596|
| 9.819190979003906|      1|   2090|
|10.030275344848633|      1|   2087|
| 11.02070426940918|      1|   2394|
|11.608592987060547|      1|   2096|
|    9.557373046875|      1|   2081|
|  9.46037769317627|      1|   1025|
| 9.815552711486816|      1|   2384|
|10.764541625976562|      1|     13|
| 10.82581615447998|      1|   2083|
| 9.720121383666992|      1|   1032|
| 10.39176082611084|      1|    551|
|10.471624374389648|      1|   3213|
|10.514969825744629|      1|   3159|
|12.995258331298828|      1|    364|
|11.679502487182617|      1|   1688|
+------------------+-------+-------+
only showing top 20 rows



### Calculating metrics

In [28]:
init_args = {"query_column": "user_id", "rating_column": "score"}

In [29]:
result_metrics = OfflineMetrics(
    [Recall(TOPK), Precision(TOPK), MAP(TOPK), NDCG(TOPK), MRR(TOPK), HitRate(TOPK)], **init_args
)(recommendations.toPandas(), raw_test_gt)

In [30]:
metrics_to_df(result_metrics)

k               1         10        20        100
HitRate    0.039238  0.218377  0.313742  0.587086
MAP        0.039238  0.082451  0.088997  0.095790
MRR        0.039238  0.082451  0.088997  0.095790
NDCG       0.039238  0.113971  0.137991  0.188293
Precision  0.039238  0.021838  0.015687  0.005871
Recall     0.039238  0.218377  0.313742  0.587086

### User embeddings

Got 6040 x 300 user embeddings, because among all 12 batches: 

11 batches contains 512 samples

1 batch contains 408 left samples

11 * 512 + 408 == 6040

In [34]:
user_embeddings

tensor([[ 1.4416,  5.4916, -1.2544,  ..., -1.3986, -5.5016,  5.2256],
        [ 3.0259, -4.0708,  1.8471,  ...,  3.0565, -0.5209, -3.2557],
        [ 0.8891, -2.1141, -0.8330,  ..., -1.4599,  3.0288, -0.7955],
        ...,
        [-1.1561,  1.6767,  0.7947,  ..., -1.8755,  0.7974, -0.3798],
        [-2.6349,  1.3446,  1.3022,  ...,  0.2333, -1.2943, -2.1294],
        [ 1.3409, -0.5924,  2.1425,  ..., -4.9075, -1.0480,  0.6077]],
       device='cuda:0')

In [35]:
user_embeddings.shape

torch.Size([6040, 300])

You can access user embeddings directly with `Bert4RecModel` class

In [39]:
device = "cuda" if torch.cuda.is_available() else "cpu"

core_model = Bert4RecModel(
    tensor_schema,
    num_blocks=2,
    num_heads=4,
    max_len=MAX_SEQ_LEN,
    hidden_size=300,
    dropout=0.5
)
core_model.eval()
core_model = core_model.to(device)

# Get first batch of data 
data = next(iter(prediction_dataloader))
tensor_map, padding_mask, tokens_mask = data.features, data.padding_mask, data.tokens_mask

# Ensure everything is on the same device
padding_mask = padding_mask.to(device)
tokens_mask = tokens_mask.to(device)
tensor_map["item_id_seq"] = tensor_map["item_id_seq"].to(device)

# Get user embeddings
user_embeddings_batch = core_model.get_query_embeddings(tensor_map, padding_mask, tokens_mask)
user_embeddings_batch

tensor([[ 0.2140, -0.2030, -0.4121,  ...,  0.0073, -0.6272, -0.6086],
        [ 0.3871, -0.0352,  0.6277,  ..., -0.1189,  0.2880, -0.2076],
        [ 0.4881,  0.4000, -0.1426,  ..., -0.0315, -0.4761,  0.1595],
        ...,
        [ 0.4864, -0.5245,  0.8303,  ..., -0.2758,  0.3587,  0.0114],
        [ 1.0421,  0.3794,  0.0295,  ...,  1.0383, -1.2656, -0.1567],
        [ 1.1374,  0.9594, -0.2814,  ...,  0.4843, -1.0908,  0.4050]],
       device='cuda:0', grad_fn=<SliceBackward0>)

In [40]:
user_embeddings_batch.shape

torch.Size([512, 300])

## Example of launching an inference for a single user without using a trainer (in order to speed up)
An example for the production of an online script

Let's assume that the user's sequence consisted of a sequence of items [1, 2, 3, 4, 5]. 
Сreate a padding mask and tokens mask corresponding to the sequence of items.

It is important to take only the latest MAX_SEQ_LEN or less items.

In [41]:
item_sequence = torch.arange(1, 5).unsqueeze(0)[:, -MAX_SEQ_LEN:]
padding_mask = torch.ones_like(item_sequence, dtype=torch.bool)
tokens_mask = padding_mask.roll(-1, dims=0)
tokens_mask[-1, ...] = 0
sequence_item_count = item_sequence.shape[1]

### Wrapping created tensors in the Bert4RecPredictionBatch entity

In [42]:
batch = Bert4RecPredictionBatch(
    query_id=torch.arange(0, item_sequence.shape[0], 1).long(),
    padding_mask=padding_mask,
    features={ITEM_FEATURE_NAME: item_sequence.long()},
    tokens_mask=tokens_mask
)

### Run predict step of the Bert4Rec and get scores from the model

In [44]:
with torch.no_grad():
    scores = best_model.predict_step(batch, 0)
scores

tensor([[ 1.2738, -2.2602, -2.4750,  ..., -4.2028, -3.5254, -2.1626]])

### Getting five items with the highest score

In [45]:
torch.topk(scores, k=5).indices

tensor([[ 847, 1203, 1959, 1284, 1214]])